lib/segment/src/index/payload_config.rs

```diff
@@ -34,3 +34,106 @@ impl PayloadConfig {
         Ok(atomic_save_json(path, self)?)
     }
 }
+
+#[derive(Debug, Deserialize, Serialize, Clone, PartialEq)]
+pub struct PayloadFieldSchemaWithIndexType {
+    #[serde(flatten)]
+    pub schema: PayloadFieldSchema,
+
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub index_types: Vec<FullPayloadIndexType>,
+}
+
+impl PayloadFieldSchemaWithIndexType {
+    pub fn new(schema: PayloadFieldSchema, index_types: Vec<FullPayloadIndexType>) -> Self {
+        Self {
+            schema,
+            index_types,
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize, Clone, PartialEq)]
+pub enum FullPayloadIndexType {
+    IntIndex(IndexMutability),
+    DatetimeIndex(IndexMutability),
+    IntMapIndex(IndexMutability),
+    KeywordIndex(IndexMutability),
+    FloatIndex(IndexMutability),
+    GeoIndex(IndexMutability),
+    FullTextIndex(IndexMutability),
+    BoolIndex(IndexMutability),
+    UuidIndex(IndexMutability),
+    UuidMapIndex(IndexMutability),
+    NullIndex(IndexMutability),
+}
+
+impl FullPayloadIndexType {
+    pub fn mutability(&self) -> &IndexMutability {
+        match self {
+            FullPayloadIndexType::IntIndex(index_mutability)
+            | FullPayloadIndexType::DatetimeIndex(index_mutability)
+            | FullPayloadIndexType::IntMapIndex(index_mutability)
+            | FullPayloadIndexType::KeywordIndex(index_mutability)
+            | FullPayloadIndexType::FloatIndex(index_mutability)
+            | FullPayloadIndexType::GeoIndex(index_mutability)
+            | FullPayloadIndexType::FullTextIndex(index_mutability)
+            | FullPayloadIndexType::BoolIndex(index_mutability)
+            | FullPayloadIndexType::UuidIndex(index_mutability)
+            | FullPayloadIndexType::UuidMapIndex(index_mutability)
+            | FullPayloadIndexType::NullIndex(index_mutability) => index_mutability,
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq)]
+pub enum IndexMutability {
+    Mutable(StorageType),
+    Immutable(StorageType),
+    Mmap { is_on_disk: bool },
+}
```

Comment:
I don't think `Mmap` variant belongs here. Correct me if I am wrong, but it should be possible to represent mmap indices as mutable or immutable with a mmap storage, isn't it?

```suggestion
pub enum IndexMutability {
    /// For appendable segments. Supports appending and deleting values
    Mutable(StorageType),
    /// For non-appendable segments. Only supports deleting values.
    Immutable(StorageType),
}
```

This should be enough for mutability, I believe. 

Also, I'm 99% sure there are no immutable indices which don't support deletions. Can anyone confirm?